sys/net/application_layer/nanocoap/sock.c

```diff
@@ -102,41 +136,100 @@ ssize_t nanocoap_sock_request(nanocoap_sock_t *sock, coap_pkt_t *pkt, size_t len
                 continue;
             }
 
-            break;
+            switch (coap_get_type(pkt)) {
+            case COAP_TYPE_RST:
+                /* TODO: handle different? */
+                return -EBADMSG;
+            case COAP_TYPE_CON:
+                /* TODO: send ACK response */
+                /* fall-through */
+            case COAP_TYPE_NON:
+                if (pkt->payload_len == 0) {
+                    return 0;
+                }
+                /* fall-through */
+            case COAP_TYPE_ACK:
+                if (pkt->payload_len == 0 && response) {
+                    state = STATE_AWAIT_RESPONSE;
+                    /* TODO: adjust timeout */
+                    break;
+                }
+                /* call user callback */
+                res = cb(arg, pkt);
+                break;
+            }
         }
     }
 
     return res;
 }
 
+static int _request_cb(void *arg, coap_pkt_t *pkt)
+{
+    struct iovec *buf = arg;
+    size_t pkt_len = coap_get_total_len(pkt);
+
+    if (pkt_len > buf->iov_len) {
+        return -ENOBUFS;
+    }
+
+    memcpy(buf->iov_base, pkt->hdr, pkt_len);
+
+    pkt->hdr = buf->iov_base;
+    pkt->token  = (uint8_t*)pkt->hdr + sizeof(coap_hdr_t);
+    pkt->payload = (uint8_t*)pkt->hdr + (pkt_len - pkt->payload_len);
+
+    return pkt_len;
+}
+
+ssize_t nanocoap_sock_request(sock_udp_t *sock, coap_pkt_t *pkt, size_t len)
```

Comment:
I'm aware that this function is preexisting, but isn't the length already implicitly known due to the data provided by `pkt`? This looks to me like inconsistent interpretations of the pkt length could be introduced via the `len` parameter, if the caller doesn't keep `pkt->payload_len` in sync.

I have the fear that this can result in fun bugs :/

Reply:
> isn't the length already implicitly known due to the data provided by `pkt`?

No `pkt` only contains the length of the request, but the idea is to re-use the request buffer for the response - whose length needs to be supplied separately.

(Yea I know… a separate buffer, argument that could just so happen to be an alias for the pkg buffer would be the cleaner solution here.)